### DISCUSSION TITLE
Mistral Mixtral 8x7b very slow

### DISCUSSION THREAD

**Discussion by thosch-lab** [oobabooga/text-generation-webui#5101] [Q&A]
Hi everybody, 

my machine: 

CPU: 14900k
GPU: 4090 RTX
RAM: 96 GB
SSD: 2 TB NVMe with 5 GB / s

so when I load my model: 

it generate Output in 151.62 seconds with 0.35 tokens/s

but the model loads fast ... in 35 sec... 

I don't know - what can I do for more tokens ... 

copy from terminal ... 

17:10:36-383111 INFO     Loading mistralai_Mixtral-8x7B-v0.1
Loading checkpoint shards: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 19/19 [00:38<00:00,  2.04s/it]
17:11:17-888910 INFO     LOADER: Transformers
17:11:17-888910 INFO     TRUNCATION LENGTH: 32768
17:11:17-889910 INFO     INSTRUCTION TEMPLATE: Alpaca
17:11:17-889910 INFO     Loaded the model in 41.51 seconds.
Output generated in 151.62 seconds (0.35 tokens/s, 53 tokens, context 137, seed 1873878655)

![image](https://github.com/oobabooga/text-generation-webui/assets/81586441/adb1979d-9670-4006-855f-eba91ea4fd63)

any idea ?

best regards

**Comment by michusx** [comment]
It looks that you are trying to use unquantized version of Mixtral. "Full" version requires a lot of memory because it need to load all 8 models each 7B parameters (even it uses only two best for inference of input). So, in your case even very powerful RTX4090 uses shared system memory for processing of Mixtral, because 24GB of VRAM is definitely not enough. And that is a bottleneck. Get a GPTQ version of Mixtral (or similar quantized one) and it will run very fast on your setup...
PS.
4bit Mixtral (like GPTQ version) still barely fits 24GB so it might not be easy. Also it works only with newest version of Transformers and AutoGPTQ - check its corresponding Hughingface page for details related to chosen model version.

**Comment by gerroon** [comment]
I have mistralai_Mistral-7B-Instruct-v0.1  and it is still slow for a super simple question. I have an RTX 2070 , 64GB ram

```
output generated in 59.68 seconds (0.42 tokens/s, 25 tokens, context

```

**Reply by TheDarkTrumpet** [reply]
It's going to be slow if you're using CPU, that's the real problem here.  Ideally you want your models to fit entirely in VRAM and use the GPU if at all possible.

There are some workarounds that can increase speed, but I haven't found good options in text-generation-webui.  Kobold can use some efficiencies that I haven't found  here, see https://github.com/LostRuins/koboldcpp - but it takes work to setup, and it's still quite slow.

**Reply by gerroon** [reply]
I am using GPU, I see Python.exe is using %100 of my GPU, and about %10 of my cores. I installed the NVDIA version during the setup, which I use natively.

I actually use the same model under Ollama and to be honest it is like 5x faster, and I use that under WSL

**Reply by TheDarkTrumpet** [reply]
Makes sense, so it appears you're also going across cpu/gpu, how many layers did you pump into the GPU?

Also, if you're using Linux (unsure on Windows on this), if you run `nvidia-smi`, how much memory do you see consumed by the main process?

**Comment by TheDarkTrumpet** [comment]
@michusx's answer is quite good, I want to add in a bit more.

Along with reducing the quantization of the model you want to load (thus reducing the memory requirement), you also have things to contend with such as that context size.  32k with a model this big is going to be a real problem.  Check https://huggingface.co/TheBloke/MixtralOrochi8x7B-GGUF under the 'Provided Files' section to see the memory requirements.

One thing you could try (although I've had _really_ bad luck with this feature) is to offload as many layers to the GPU as possible - and keep the rest on the CPU.  Alternatively, I've heard good things about AWQ models taking less RAM than other models - although I haven't tried this.  But, you can offload as much as possible onto your GPU, which may help.  Ingestion of tokens is the slowest part I've seen with models that splice between the two, so that may be worth trying as well.

**Reply by michusx** [reply]
When I saw question about speed-up inference (especially on machine with 4090) then my first thought was how to get double digit token/s value, not change from 0.35 to ~2... :-) And that's why I focus on fitting the model in VRAM as allowing to traverse between RAM and VRAM reduce speed dramatically - from my observation - no matter what method you apply.
So, if improvement to aforementioned 2 tokens/s is on table then of course everything you mentioned apply @TheDarkTrumpet
32K context is additional killer so it makes more VRAM memory requirement. TheBloke at Huggingface published 3bit GPTQ that requires about 18G of memory so there will be more left for context (on 4090). But I don't know what is the loss of quality of this 3bit quantization. Probably soon Exllama2 provide support for Mixtral and EXL2 format could be used with good size VS quality ratio...

**Reply by TheDarkTrumpet** [reply]
Yeah, if your goal is to get really high inference, then you could work with that lower bit models with a lower context size.  

In terms of how it all works, https://deci.ai/quantization-and-quantization-aware-training/ has some good information on it.  I have noticed differences between 4 and 5 bit quantizations with models I commonly use.  How significant they are is really a question, and it could have been problems in the quantization method. There's some extreme cases I found where the model wasn't really useful on the lower bits (for my purposes).

I'd try with as large as you can handle with the context size you want to accomplish.  if your goal is to deal with very large context sizes, you may need to look at a method using something like langchain (or SillyTavern has extras that do this for you automatically) to summarize portions of the old context so that you can feed it back in.  Really depends on what you want to do with the model.  Just remember that the context window includes the prompt + output as a total amount, and counts as what the model can handle.  So in reality, you're sending: Prompt + context + expected max tokens = context window.

**Reply by michusx** [reply]
I'm just in prcess of studying and testing all these RAG methodology. I my cese it is scenario for working with technical documentation libraries where even 32k or 200k  context would not be sufficient 😉 (BTW, I found challenging the process of extraction of different datatypes (especially tables) from sources like PDFs, HTMLs etc. And summarization can be done even with simpler models like regular Mistral or Llama2 7/13B).
I faced large context memory requirements problem using nice model: Phind-Codellama 34B for coding support... Even on PV100 with 32GB of VRAM and 4bit GPTQ version there is not so much memory left for large context 😀
So, coming back to the original post of author of this topic. We don't know what is his usecase but I think we've provided him some ideas for individual research... 😁

**Reply by TheDarkTrumpet** [reply]
If you have a chance, take a look at LangChain.  There's a map function that can break things up into smaller chunks and process independently in a model, and then combine the results after the fact.  It's been a recent fascination of mine, along with the ability to call out to external services, or other models, etc.  

I use the same Phind-Codellama 34B model as well.  Luckily, I have the architecture to allow it with very large context (16k to 32k).  Unsure if you're a mac guy, but one option besides one of the 48Gb cards is to look at the M1 or M2 macs with unified memory.  The cool thing about those systems is that the "video ram" is seen as the same as "system ram".  So you can load some pretty high context models and it works quite well.  I run the Phind-Codellama on here mostly, and use it for all sorts of stuff.  A basic general purpose model, including for coding.  I have 64Gb of RAM on my M2, and it really meets that need.

**Reply by michusx** [reply]
Yes, I'm familiar with Langchain - of course rather its basics but managed to build simple RAG with oobabooga/textgen connected by OpenAPI working with Phind-Codellama for inference.
And playing with that "system" I faced a lot of challenges with tables extraction (as technical docs are full of different tables). "Unstructured" library looks promising for text but still tables are separate topic. So step by step I'm delving deep into advanced RAG techniques...
I'm working on Quadro GV100 in my company (and 4060Ti 16GB at home) but heard a lot of good about Apple M elasticity in LLM area...
Definitely all those LLM things are developing fast and we are at rather begining phase with all its limitations.
